### ISSUE TITLE
The API client does not have default timeouts set

### ISSUE THREAD

**Issue by asurak** [algolia/algoliasearch-client-python#119]
The default connect timeout is unknown and the client breaks the required specification for explicit timeouts.

**Comment by ercolanelli-leo** [comment]
@asurak the timeouts are passed down to the underlying library during the HTTP call (https://github.com/algolia/algoliasearch-client-python/blob/master/algoliasearch/transport.py#L73-L75), how did you diagnose the issue?

**Comment by asurak** [comment]
They are passed but what is the default value if the user does not explicitly set it?

**Comment by ercolanelli-leo** [comment]
Here: https://github.com/algolia/algoliasearch-client-python/blob/master/algoliasearch/transport.py#L31-L32
The connection timeout is 2 seconds, the read timeouts depends on whether we are doing a search or an indexation: 5 seconds for search, 10 seconds for indexation

**Comment by kylegibson** [comment]
We're experiencing this timeout issue right now with DNS resolution. It's taking the python client 20 seconds to switch to the next host. I believe it has something to do with requests retrying. 

```
algolia = AlgoliaIndex(algolia_hosts=['BxxxxxxxR-dsn.algolia.net'])
In [10]: print datetime.now(); r = algolia.index.search('test'); 
2016-12-01 15:22:37.276561

---------------------------------------------------------------------------
AlgoliaException                          Traceback (most recent call last)
<ipython-input-10-05310073e668> in <module>()
----> 1 print datetime.now(); r = algolia.index.search('test');

/home/policystat/env/lib/python2.7/site-packages/algoliasearch/index.pyc in search(self, query, args)
    397         args['query'] = query
    398         params = {'params': urlencode(urlify(args))}
--> 399         return self._req(True, '/query', 'POST', data=params)
    400 
    401     @deprecated

/home/policystat/env/lib/python2.7/site-packages/algoliasearch/index.pyc in _req(self, is_search, path, meth, params, data)
    887         """Perform an HTTPS request with retry logic."""
    888         path = '%s%s' % (self._request_path, path)
--> 889         return self.client._req(is_search, path, meth, params, data)

/home/policystat/env/lib/python2.7/site-packages/algoliasearch/client.pyc in _req(self, is_search, path, meth, params, data)
    500                 data = {}
    501             data['apiKey'] = self.api_key
--> 502         return self._transport.req(is_search, path, meth, params, data)

/home/policystat/env/lib/python2.7/site-packages/algoliasearch/transport.pyc in req(self, is_search, path, meth, params, data)
    123                 exceptions[host] = "%s: %s" % (e.__class__.__name__, str(e))
    124 
--> 125         raise AlgoliaException('Unreachable hosts: %s' % exceptions)

AlgoliaException: Unreachable hosts: {'B2O5OZ3T4R-dsn.algolia.net': "ConnectionError: HTTPSConnectionPool(host='BxxxxxxxR-dsn.algolia.net', port=443): Max retries exceeded with url: /1/indexes/documents_local_base/query (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x61421d0>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))"}

In [11]: print datetime.now();
2016-12-01 15:22:57.317488

```

**Comment by ercolanelli-leo** [comment]
Hi @kylegibson,

It's really unfortunate you are experiencing such an issue. I would also think the retry logic is the culprit, but the fact that you are specifying only 1 host imply that there won't be any retry after the first failure (see here https://github.com/algolia/algoliasearch-client-python/blob/master/algoliasearch/transport.py#L110-L123)
However there is something wrong with the ever increasing timeout that I intend to fix ASAP.

**Comment by kylegibson** [comment]
> the fact that you are specifying only 1 host imply that there won't be any retry after the first failure

I am referring to request's internal retry logic, not Algolia's. See the exception that is raised above "Max retries exceeded with url".

The reason I specified only one host is otherwise it fails over to the next host and succeeds, and exception raised by the first host is hidden.

**Comment by ercolanelli-leo** [comment]
Good catch! That seems to be a plausible explanation.
I just did a few try using `session.mount('https://', HTTPAdapter(max_retries=1)'` but I don't see any improvement unfortunately. I will continue to try things out.

**Comment by ercolanelli-leo** [comment]
As a temporary workaround, you can put `['<APPID>-1.algolianet.com','<APPID>-2.algolianet.com','<APPID>-3.algolianet.com']` in the hosts, or change your `/etc/host` by adding the line `<APPID>-1.algolianet.om <APPID>-dsn.algolia.net`

**Comment by kylegibson** [comment]
> or change your /etc/host by adding the line <APPID>-1.algolianet.om <APPID>-dsn.algolia.net

This is the workaround that we implemented a few hours ago, and also a few weeks ago.

**Comment by ercolanelli-leo** [comment]
@kylegibson I have news (and good one), to improve the situation.

- The DNS lookup done by `urllib3` uses `getaddrinfo` which uses timeouts and retries directly from `/etc/resolv.conf`. The good news is that those values can be read from the environment. I will push a PR tomorrow that sets those values in the process environment. this will avoid this 20 seconds delay, and instead drop at about 2

- I will implement a timer to blacklist the host for several minutes after an unsuccessful requests. This will avoid having this timeout on *all* the queries, but instead only every few minutes.

What do you think?

**Comment by ercolanelli-leo** [comment]
PR here https://github.com/algolia/algoliasearch-client-python/pull/123

**Comment by ercolanelli-leo** [comment]
The fix is published in version 1.11.0 of the client, available on pypi. I am closing this issue for now.
